### DISCUSSION TITLE
Sony LDAC bluetooth codec support on mac

### DISCUSSION THREAD

**Discussion by p6002** [Homebrew/discussions#2606] [Tap maintenance and Homebrew development]
Since there is support in Androd and Linux, is it possible to add LDAC support as homebrew to macOS/hackintosh?
https://en.wikipedia.org/wiki/LDAC_(codec)
Many people with expensive Sony headphones must use the low quality SBC / AAC codec on their mac and would be happy to be able to use LDAC instead.
https://aur.archlinux.org/packages/pulseaudio-modules-bt/
https://github.com/EHfive/pulseaudio-modules-bt
https://eischmann.wordpress.com/2019/02/11/better-bluetooth-sound-quality-on-linux/
https://ap.community.sony.com/s/question/0D54O00006PvV9tSAF/ldac-bluetooth-usb-transmitter?language=en_US

**Comment by SMillerDev** [comment]
There is a pulseaudio formula so you could compile modules for that I think. If it needs kernel support you're out of luck though.

**Reply by mcnaugha** [reply]
pulseaudio 15.0 adds support for LDAC and AptX HD. The bottled version is currently 14.2. How do we get it updated to 15.0 or is there a way we can build 15.0 under homebrew for macOS?

15.0 is a significant release. With the integrated support for LDAC and AptX HD there is a possibility to support Bluetooth Audio receivers (e.g. LDAC-enabled headphones) that support these codecs and output higher bit rate audio from Apple Music to such receivers. This would be a major first for the Mac. 

Cannot believe how long Apple is taking to deliver its own solution to the problem. Qualcomm just announced AptX Lossless codec. Apple has to have a trick up its sleeve. I'd like to imagine that it's to say goodbye Bluetooth Audio and hello Wi-Fi audio or something so that there is no need for new codecs or lossy formats.

**Reply by gromgit** [reply]
> How do we get it updated to 15.0

Someone has to figure out how to solve the build failures in [this 5-month-old PR](https://github.com/Homebrew/homebrew-core/pull/82113).

**Reply by mcnaugha** [reply]
What's the deal with Pipewire then? Has it been tried and it didn't work too?

I got access to LDAC via Fedora 35 on Parallels on M1 Pro Mac, if that gives any hints as to what components we need ported.

Thanks
A.

**Reply by alebcay** [reply]
I've taken a look at the CI failures in that linked PR.
- The Apple Silicon-specific failures are related to some CPU features/info detection headers on macOS that should not be included in the build (they're Intel-specific). Pulseaudio source will need some adjustment to conditionally include the problematic header. (I suspect there may be more Apple Silicon-specific fixes needed after moving past this one)
- Some of the Unix socket code in Pulseaudio assumes existence of `ucred` or `cmsgcred` structs (for Linux/FreeBSD/GNU Hurd), but macOS only has `xucred`, which may or may not be suitable a replacement. There would need to be work done on porting or abstracting any code that works with `ucred`/`cmsgcred` (mostly in `src/iochannel.c` it seems) to also support `xucred` or some other mechanism that works on macOS.

I don't think we've tried anything with Pipewire here at Homebrew (it's not shipped in our repository). Based on some rudimentary googling, it sounds like there has not been much impetus on making it work (or even seeing how much work might be needed) on macOS.

**Reply by mcnaugha** [reply]
Which is unfortunate because enabling AptX HD and LDAC on macOS would make you instant heroes to audio nerds. The Internet might explode as they say in pop culture.

It's become interesting since Apple Music launched its lossless and Hi-Res service last Summer. Every Mac and iPhone/iPad user is disadvantaged while these encoders are not available. To be fair, it should have been interesting with any previously available lossless format too.
